tests/unit/components/test_braveadblock.py

```diff
@@ -0,0 +1,320 @@
+        "https://damndelicious.net/2019/04/03/easy-slow-cooker-chili/",
+        ResourceType.image,
+    ),
+]
+
+
+def run_function_on_dataset(given_function):
+    """Run the given function on a bunch of urls.
+
+    In the data folder, we have a file called `adblock_dataset.tsv`, which
+    contains tuples of (url, source_url, type) in each line. We give these
+    to values to the given function, row by row.
+    """
+
+    def dataset_type_to_enum(type_int: int) -> ResourceType:
+        """Translate the dataset's encoding of a resource type to Qutebrowser's."""
+        if type_int == 0:
+            return ResourceType.unknown
+        elif type_int == 1:
+            return ResourceType.image
+        elif type_int == 2:
+            return ResourceType.stylesheet
+        elif type_int == 3:
+            return ResourceType.media
+        elif type_int == 4:
+            return ResourceType.script
+        elif type_int == 5:
+            return ResourceType.font_resource
+        elif type_int == 6:
+            return ResourceType.xhr
+        else:
+            assert type_int == 7
+            return ResourceType.sub_frame
+
+    dataset_path = os.path.join(THIS_DIR, "data", "adblock_dataset.tsv")
+    with open(dataset_path, "r", encoding="utf-8") as f:
+        reader = csv.DictReader(f, delimiter="\t")
+        for row in reader:
+            url = QUrl(row["url"])
+            source_url = QUrl(row["source_url"])
+            resource_type = dataset_type_to_enum(int(row["type"]))
+            given_function(url, source_url, resource_type)
+
+
+def assert_none_blocked(ad_blocker):
+    assert_urls(ad_blocker, NOT_OKAY_URLS + OKAY_URLS, False)
+
+    def assert_not_blocked(ad_blocker, url, source_url, resource_type):
+        assert not ad_blocker._is_blocked(url, source_url, resource_type)
+
+    run_function_on_dataset(
+        lambda url, source_url, resource_type: assert_not_blocked(
+            ad_blocker, url, source_url, resource_type
+        )
+    )
+
+
+def create_blocklist_invalid_utf8(directory) -> str:
+    dest_path = os.path.join(directory, "invalid_utf8.txt")
+    with open(dest_path, "wb") as f:
+        f.write(b"invalidutf8\xa0")
+    return QUrl.fromLocalFile(dest_path).toString()
+
+
+def create_easylist_easyprivacy(directory) -> List[str]:
+    """Copy the easyprivacy and easylist blocklists into the given dir."""
+    urls = []
+    for blocklist in ["easyprivacy.txt", "easylist.txt"]:
+        bl_src_path = os.path.join(THIS_DIR, "data", blocklist)
+        bl_dst_path = os.path.join(directory, blocklist)
+        copy(bl_src_path, bl_dst_path)
+        assert os.path.isfile(bl_dst_path)
+        urls.append(QUrl.fromLocalFile(bl_dst_path).toString())
+    return urls
+
+
+@pytest.fixture
+def ad_blocker_factory(config_tmpdir, data_tmpdir, download_stub, config_stub):
+    def factory():
+        return BraveAdBlocker(engine_factory=lambda: Engine([]), data_dir=data_tmpdir)
```

Comment:
Why does this need to be a factory function rather than just returning the adblocker?